lib/bundle.js

```diff
@@ -76,6 +77,101 @@ function extractBundleName(fileName) {
   return path.relative(config.pjson.baseURL, fileName.replace(/\.js$/, '')).replace(/\\/g, '/');
 }
 
+function getModuleName (file) {
+  var baseURL =  config.pjson.baseURL + '/';
+  var paths =  systemBuilder.loader.paths;
+
+  //todo: cache the sortedPaths for subsequent usage for better performance.
+  var sortedPaths = sortPathsByValue(paths); // sort the rules in descending order so that the longest specific rule gets hoisted up top.
+
+  var id = '';
+  for (p in sortedPaths) {
+    var rule = sortedPaths[p];
+    var resolved = path.resolve(baseURL, rule); // resolve the full path from the rule
+    resolved = resolved.replace(/\\/g, '/'); // replace  '\' with  '/'
+
+    var ruleDir =  path.dirname(resolved); // find the resolved directory
+    var fileDir = path.dirname(file);      // find the file's directory.
+
+    var regex = new RegExp('^' + ruleDir, 'i'); // test if wildcard's directory path has a match with file's directory.
+
+    if(regex.test(fileDir)){
+      var parts = p.split('*');
+      var moduleName =  file.replace(regex, '');
+      moduleName = moduleName.replace(/^\//, ''); //remove the last trailing '/' from the baseURL
+      moduleName =  moduleName.replace(new RegExp(path.extname(rule) + '$', 'g'), ''); // remove any extension the rule have from the moduleName.
+      id = parts[0] + moduleName;
+      break;
+    }
+  }
+
+  if(id === ''){
+    throw new Error('No matching rule defined for file: ' + "'" + file + "'");
+  }
+
+  return id;
+}
+
+function sortPathsByValue(obj) {
+  // convert object into array
+  var sortable = [];
+  for (var key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      sortable.push([key, obj[key]]);
+    }
+  }
+  // sort items by path value in descending order
+  sortable.sort(function(a, b) {
+    var x = a[1].toLowerCase();
+    var y = b[1].toLowerCase();
+
+    if (x.length > y.length) return -1;
+    if (x.length < y.length) return 1;
+    return 0;
+  });
+
+  var sorted = {};
+  sortable.forEach(function(arr) {
+    sorted[arr[0]] = arr[1];
+  });
+
+  return sorted;
+};
+
+function expandGlob(moduleName) {
+  var loader = systemBuilder.loader;
+
+  return loader.normalize(moduleName)
+    .then(function (normalized) {
+      return loader.locate({name: normalized})
+    })
+    .then(function (address) {
+
+      var normalizedAddress = path.normalize(address).replace(new RegExp('\\' + path.sep, 'g'), '/'); // glob does not like `\`
+      normalizedAddress = normalizedAddress.replace(new RegExp('^file:'), '');
+
+      var files = glob.sync(normalizedAddress, {});
```

Comment:
When we do the glob, can we add glob options to remove all globbing features we're not supporting (https://github.com/isaacs/node-glob#options). I'm just worried about things like `[` or `@` in module names causing issues.